File: leetcode/sudokucheck/sudokucheck.py
class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        # check rows
        for row in board:
            numset = set()
            for col in row:
                if col == '.':
                    continue
                elif col.isdigit():
                    if col in numset:
                        return False
                    else:
                        numset.add(col)
                else:
                    return False
        
        
        # check columns
        for col in range(0, 9):
            numset = set()
            for row in board:
                letter = row[col]
                if letter == '.':
                    continue
                elif letter.isdigit():
                    if letter in numset:
                        return False
                    else:
                        numset.add(letter)
                else:
                    return False
        
        # check boxes
        for rowidx in range(0, 9):
            row = board[rowidx]
            if rowidx % 3 == 0:
                boxset = dict()
                boxset[0] = set()
                boxset[1] = set()
                boxset[2] = set()
            for colidx in range(0, 9):
                box = int(colidx / 3)
                letter = row[colidx]
                if letter == '.':
                    continue
                elif letter.isdigit():
                    if letter in boxset[box]:
                        return False
                    else:
                        boxset[box].add(letter)
                        # print(rowidx, colidx, box, letter, boxset)
                else:
                    return False
                    
        return True

File: leetcode/sudokucheck/sudokucheck.c
bool isValidSudoku(char** board, int boardSize, int* boardColSize){
    bool *numchk = (bool *)malloc(boardSize*sizeof(bool));
    bool *boxchk[3];
    
    // check rows
    for (int i = 0; i < boardSize; i++) {
        for (int j = 0; j < *boardColSize; j++) {
            numchk[j] = false;
        }
        for (int j = 0; j < *boardColSize; j++) {
            char let = board[i][j];
            if (let == '.')
                continue;
            else if (isdigit(let)) {
                int num = (int)(let - '0') - 1;
                if (numchk[num]) {
                    return false;
                } else {
                    numchk[num] = true;
                }
            } else {
                return false;
            }
        }
    }
    
    
    // check cols
    for (int j = 0; j < *boardColSize; j++) {
        for (int i = 0; i < boardSize; i++) {
            numchk[i] = false;
        }
        for (int i = 0; i < boardSize; i++) {
            char let = board[i][j];
            if (let == '.')
                continue;
            else if (isdigit(let)) {
                int num = (int)(let - '0') - 1;
                if (numchk[num]) {
                    return false;
                } else {
                    numchk[num] = true;
                }
            } else {
                return false;
            }
        }
    }
    

    // check boxes
    for (int box = 0; box < 3; box++)
        boxchk[box] = (bool *)malloc(boardSize*sizeof(bool));
    for (int i = 0; i < boardSize; i++) {
        if (i % 3 == 0) {
            for (int j = 0; j < *boardColSize; j++) {
                boxchk[0][j] = false;
                boxchk[1][j] = false;
                boxchk[2][j] = false;
            }
        }
        for (int j = 0; j < *boardColSize; j++) {
            int box = j/3;
            char let = board[i][j];
            if (let == '.')
                continue;
            else if (isdigit(let)) {
                int num = (int)(let - '0') - 1;
                if (boxchk[box][num]) {
                    return false;
                } else {
                    boxchk[box][num] = true;
                }
            } else {
                return false;
            }
        }
    }
    
    return true;
}

